File: core/audio/vad.go
package audio

import "math"

// VADConfig configures voice activity detection.
type VADConfig struct {
	// EnergyThreshold is the RMS energy level above which audio is considered speech.
	// Typical range: 0.01–0.05. Default: 0.02.
	EnergyThreshold float32
	// SilenceDurationMs is how long continuous silence triggers auto-stop.
	// Default: 3000 ms.
	SilenceDurationMs int
}

// DefaultVADConfig returns sensible defaults for VAD.
func DefaultVADConfig() VADConfig {
	return VADConfig{
		EnergyThreshold:   0.02,
		SilenceDurationMs: 3000,
	}
}

// IsSpeech returns true if the audio chunk contains speech based on RMS energy.
func IsSpeech(chunk []float32, cfg VADConfig) bool {
	if len(chunk) == 0 {
		return false
	}
	var sum float64
	for _, s := range chunk {
		sum += float64(s) * float64(s)
	}
	rms := math.Sqrt(sum / float64(len(chunk)))
	return float32(rms) > cfg.EnergyThreshold
}

File: core-rs/src/audio/vad.rs
use crate::audio::types::VadConfig;

/// Returns `true` if the audio chunk contains speech based on RMS energy.
///
/// Uses a simple energy-based VAD: if the root-mean-square of the samples
/// exceeds the configured threshold, the chunk is classified as speech.
pub fn is_speech(chunk: &[f32], cfg: VadConfig) -> bool {
    if chunk.is_empty() {
        return false;
    }
    let sum: f64 = chunk.iter().map(|&s| (s as f64) * (s as f64)).sum();
    let rms = (sum / chunk.len() as f64).sqrt();
    (rms as f32) > cfg.energy_threshold
}
